lib/src/solver/incompatibility_cause.dart

```diff
@@ -6,26 +6,14 @@ import 'package:pub_semver/pub_semver.dart';
 
 import '../exceptions.dart';
 import '../language_version.dart';
+import '../package_name.dart';
 import '../sdk.dart';
+import '../source/sdk.dart';
 import 'incompatibility.dart';
 
 /// The reason an [Incompatibility]'s terms are incompatible.
-abstract class IncompatibilityCause {
-  const IncompatibilityCause._();
-
-  /// The incompatibility represents the requirement that the root package
-  /// exists.
-  static const IncompatibilityCause root = _Cause('root');
-
-  /// The incompatibility represents a package's dependency.
-  static const IncompatibilityCause dependency = _Cause('dependency');
-
-  /// The incompatibility indicates that the package has no versions that match
-  /// the given constraint.
-  static const IncompatibilityCause noVersions = _Cause('no versions');
-
-  /// The incompatibility indicates that the package has an unknown source.
-  static const IncompatibilityCause unknownSource = _Cause('unknown source');
+sealed class IncompatibilityCause {
+  IncompatibilityCause();
 
```

Comment:
We could still do:
```suggestion
  static const IncompatibilityCause root = RootIncompatibilityCause();
```

Then we won't need to refactor the code, and we won't need to create instances for must of them, we can just keep these singletons.

Sure, `DependencyIncompatibilityCause` will have to be special.

Reply:
I like the checks of the cause to be type-checks anyway instead of equality checks. Potentially there are places where we can have exhaustive switches on the causes.

If we want to keep a singleton around... maybe... I think it is an insignificant optimization in this case.